### PR TITLE
feat(tasks): supporting scopes in tasks TT-114

### PR DESCRIPTION
**Pull request by str-anger**
Based on: https://riverlane.atlassian.net/wiki/spaces/Tooling/pages/3319070744/ERD+Job+Queuing+--+Aqueduct+Pro+scopes

🟢 ERD1.1-1B
- in `get_all_tasks` was originally supported

🟢 ERD1.1-2B
- in `get_all_tasks` was originally supported

🟢 ERD1.3-1B
- added to `orm.py` and `extensions.py`

⚪ ERD1.3-2F - TBD
⚪ ERD1.3-3F - TBD
🟢 ERD2.1-1B 
- `aqueductcore/backend/context.py`

⚪ ERD2.1-2B - this is a Core changes.
🟢 ERD2.1-3F 
- `aqueductcore/frontend/src/__mocks__/queries/user/getUserInformation.tsx`
- `aqueductcore/frontend/src/helper/auth/userScope.tsx`

⚪ ERD2.1-4F - Pro changes here
🟢 ERD2.2-1B
- added to `get_all_tasks`

🟢 ERD2.2-2B
- added to `get_all_tasks`

🟢 ERD2.2-3B
- added to `get_all_tasks`

🟢 ERD2.2-4B
- added to `get_all_tasks`

🟢 ERD2.3-2B 
- added to execute() of `extensions_executor`.

🟢 ERD2.4-1B
-  added to `revoke_task`
- 
🟢 ERD2.5-1K - ADDED to KeyCloack

### CODE REVIEW

**Review by samiralavi — APPROVED** [review]
Great work @str-anger, mostly it looked fine. I approve the PR but please check my comments and confirm there are no issues there.

File: aqueductcore/backend/services/task_executor.py [hunk 1]
```diff
@@ -215,10 +235,17 @@ async def get_all_tasks(  # pylint: disable=too-many-arguments
...
+    if not user_info.can_view_any_experiment():
         statement = statement.join(orm.Experiment).filter(
             orm.Experiment.created_by == user_info.uuid
         )
+    if not user_info.can_view_any_task():
+        statement = statement.filter(orm.Task.created_by == user_info.uuid)
+
+    # cannot view any tasks
+    if not user_info.can_view_task_owned_by(user_info.uuid):
+        # raise AQDPermission("User has no permission to view tasks")
```
[the remark this answers is not in the document]
Comment (str-anger) [reply aqueductcore/backend/services/task_executor.py:247]: Sure, will do!

File: aqueductcore/backend/services/task_executor.py [hunk 2]
```diff
@@ -91,7 +91,7 @@ async def _update_task_info(task_id: str, wait=True) -> TaskProcessExecutionResu
     """Updates information about a task. Waits until ready if asked."""
     task = AsyncResult(task_id)
     if wait:
-        if not task.ready():
+        while not task.ready():
```
[the remark this answers is not in the document]
Comment (str-anger) [reply aqueductcore/backend/services/task_executor.py:94]: I've changed default behaviour for wait=False. I'm still thinking we might benefit from blocking call in pyAqueduct. But not sure for now.

[the remark this answers is not in the document]
Comment (samiralavi) [reply aqueductcore/backend/services/task_executor.py:94]: I don't understand the relation to PyAqueduct, but if you find it helpful, sure let's keep the wait.

File: aqueductcore/backend/services/extensions_executor.py [hunk 3]
```diff
@@ -177,6 +180,19 @@ async def execute(
...
+            .options(joinedload(orm.Experiment.created_by_user))
+            .where(orm.Experiment.uuid == experiment_uuid)
+        )
+        cursor = await db_session.execute(query)
+        experiment = cursor.scalars().first()
+        if experiment is None:
+            raise AQDValidationError("Experiment not found in the database.")
+
+        if not user_info.can_create_task_in_experiment_owned_by(experiment.created_by_user.uuid):
+            raise AQDPermission("User has no permission to start tasks in this experiment.")
```
Comment (safeamiiir) [inline aqueductcore/backend/services/extensions_executor.py:194]: As we're using a different name in the Figma
```suggestion
            raise AQDPermission("User has no permission to run a job in this experiment.")
```

**Review by safeamiiir — APPROVED** [review]
I'd rather wait for Jatin to also have a look tomorrow for an other Approval, but things looks fine to me with minor comments.

**Review by samiralavi — APPROVED** [review]
The changes are fine, great work @str-anger

**Pull request merged**